File: weight_tracker.cs
// weight_tracker.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

class WeightEntry {
    public int Id { get; set; }
    public string Date { get; set; }
    public double Weight { get; set; }
    public string Notes { get; set; }
}

class WeightTracker {
    private List<WeightEntry> entries = new List<WeightEntry>();
    private const string DataFile = "weight_data.json";

    public WeightTracker() {
        Load();
    }

    private void Load() {
        if (File.Exists(DataFile)) {
            try {
                string json = File.ReadAllText(DataFile);
                entries = JsonSerializer.Deserialize<List<WeightEntry>>(json) ?? new List<WeightEntry>();
            } catch {
                entries = new List<WeightEntry>();
            }
        }
    }

    private void Save() {
        string json = JsonSerializer.Serialize(entries, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    public int Add(string date, double weight, string notes = "") {
        int id = entries.Count + 1;
        entries.Add(new WeightEntry { Id = id, Date = date, Weight = weight, Notes = notes });
        Save();
        return id;
    }

    public void ListAll() {
        if (entries.Count == 0) {
            Console.WriteLine("\u001B[33mНет записей.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[36m{"ID",-4} {"Дата",-12} {"Вес (кг)",-12} {"Заметки",-30}\u001B[0m");
        Console.WriteLine(new string('-', 60));
        foreach (var e in entries) {
            Console.WriteLine($"{e.Id,-4} {e.Date,-12} {e.Weight,-12:F1} {e.Notes,-30}");
        }
    }

    public void PrintStats() {
        if (entries.Count == 0) {
            Console.WriteLine("\u001B[33mНет данных для статистики.\u001B[0m");
            return;
        }
        var sorted = entries.OrderBy(e => e.Date).ToList();
        double first = sorted.First().Weight;
        double last = sorted.Last().Weight;
        double change = last - first;
        double avg = entries.Average(e => e.Weight);
        double min = entries.Min(e => e.Weight);
        double max = entries.Max(e => e.Weight);
        double? bmi = null;
        if (last > 0) {
            double height = 1.75;
            bmi = last / (height * height);
        }
        Console.WriteLine("\u001B[36m📊 Статистика:\u001B[0m");
        Console.WriteLine($"  Всего записей: {entries.Count}");
        Console.WriteLine($"  Текущий вес: {last:F1} кг");
        Console.WriteLine($"  Средний вес: {avg:F1} кг");
        Console.WriteLine($"  Минимальный вес: {min:F1} кг");
        Console.WriteLine($"  Максимальный вес: {max:F1} кг");
        string changeColor = change < 0 ? "\u001B[32m" : change > 0 ? "\u001B[31m" : "\u001B[33m";
        Console.WriteLine($"  Изменение: {changeColor}{change:+F1} кг\u001B[0m");
        if (bmi.HasValue) {
            string bmiColor = bmi >= 18.5 && bmi <= 24.9 ? "\u001B[32m" : bmi >= 25 && bmi <= 29.9 ? "\u001B[33m" : "\u001B[31m";
            Console.WriteLine($"  ИМТ: {bmiColor}{bmi:F1}\u001B[0m");
        }
    }

    public void ExportCSV(string filename = "weight_data.csv") {
        if (entries.Count == 0) {
            Console.WriteLine("\u001B[33mНет данных для экспорта.\u001B[0m");
            return;
        }
        using var writer = new StreamWriter(filename);
        writer.WriteLine("ID,Дата,Вес (кг),Заметки");
        foreach (var e in entries) {
            writer.WriteLine($"{e.Id},{e.Date},{e.Weight:F1},{e.Notes}");
        }
        Console.WriteLine($"\u001B[32m💾 Экспорт CSV: {filename}\u001B[0m");
    }

    public void ExportJSON(string filename = "weight_data_export.json") {
        if (entries.Count == 0) {
            Console.WriteLine("\u001B[33mНет данных для экспорта.\u001B[0m");
            return;
        }
        string json = JsonSerializer.Serialize(entries, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
        Console.WriteLine($"\u001B[32m💾 Экспорт JSON: {filename}\u001B[0m");
    }

    public bool Delete(int id) {
        var entry = entries.FirstOrDefault(e => e.Id == id);
        if (entry != null) {
            entries.Remove(entry);
            Save();
            return true;
        }
        return false;
    }

    public bool Edit(int id, string field, string value) {
        var entry = entries.FirstOrDefault(e => e.Id == id);
        if (entry == null) return false;
        switch (field.ToLower()) {
            case "date": entry.Date = value; break;
            case "weight": if (double.TryParse(value, out double w)) entry.Weight = w; else return false; break;
            case "notes": entry.Notes = value; break;
            default: return false;
        }
        Save();
        return true;
    }

    public void FilterByDate(string start, string end) {
        var results = entries.Where(e => e.Date.CompareTo(start) >= 0 && e.Date.CompareTo(end) <= 0).ToList();
        if (results.Count == 0) {
            Console.WriteLine("\u001B[33mНет записей за указанный период.\u001B[0m");
            return;
        }
        foreach (var e in results) {
            Console.WriteLine($"{e.Id}: {e.Date} | {e.Weight:F1} кг | {e.Notes}");
        }
    }

    public static void Main() {
        var tracker = new WeightTracker();
        while (true) {
            Console.WriteLine("\n\u001B[36m⚖️ Weight Tracker Pro (C#)\u001B[0m");
            Console.WriteLine("1. Добавить запись");
            Console.WriteLine("2. Показать все записи");
            Console.WriteLine("3. Статистика");
            Console.WriteLine("4. Экспорт в CSV");
            Console.WriteLine("5. Экспорт в JSON");
            Console.WriteLine("6. Удалить запись");
            Console.WriteLine("7. Редактировать запись");
            Console.WriteLine("8. Фильтр по дате");
            Console.WriteLine("9. Выход");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine();
            switch (choice) {
                case "1":
                    Console.Write("Дата (ГГГГ-ММ-ДД, Enter для сегодня): ");
                    string date = Console.ReadLine();
                    if (string.IsNullOrEmpty(date)) date = DateTime.Now.ToString("yyyy-MM-dd");
                    Console.Write("Вес (кг): ");
                    double weight = double.Parse(Console.ReadLine());
                    Console.Write("Заметки (опционально): ");
                    string notes = Console.ReadLine();
                    int id = tracker.Add(date, weight, notes);
                    Console.WriteLine($"\u001B[32m✅ Запись добавлена (ID: {id})\u001B[0m");
                    break;
                case "2": tracker.ListAll(); break;
                case "3": tracker.PrintStats(); break;
                case "4": tracker.ExportCSV(); break;
                case "5": tracker.ExportJSON(); break;
                case "6":
                    tracker.ListAll();
                    Console.Write("Введите ID для удаления: ");
                    int delId = int.Parse(Console.ReadLine());
                    if (tracker.Delete(delId)) {
                        Console.WriteLine("\u001B[32m✅ Запись удалена.\u001B[0m");
                    } else {
                        Console.WriteLine("\u001B[31m❌ Запись не найдена.\u001B[0m");
                    }
                    break;
                case "7":
                    tracker.ListAll();
                    Console.Write("Введите ID для редактирования: ");
                    int editId = int.Parse(Console.ReadLine());
                    Console.Write("Какое поле редактировать (date, weight, notes): ");
                    string field = Console.ReadLine().ToLower();
                    Console.Write("Новое значение: ");
                    string value = Console.ReadLine();
                    if (tracker.Edit(editId, field, value)) {
                        Console.WriteLine("\u001B[32m✅ Запись обновлена.\u001B[0m");
                    } else {
                        Console.WriteLine("\u001B[31m❌ Не удалось обновить.\u001B[0m");
                    }
                    break;
                case "8":
                    Console.Write("Начальная дата (ГГГГ-ММ-ДД): ");
                    string start = Console.ReadLine();
                    Console.Write("Конечная дата (ГГГГ-ММ-ДД): ");
                    string end = Console.ReadLine();
                    tracker.FilterByDate(start, end);
                    break;
                case "9":
                    Console.WriteLine("До свидания!");
                    return;
                default:
                    Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m");
                    break;
            }
        }
    }
}

File: weight_tracker.py


### 1. `weight_tracker.py` (Python)

```python
# weight_tracker.py — Python версия

import json
import os
import csv
from datetime import datetime, timedelta
from colorama import init, Fore, Style

init(autoreset=True)
DATA_FILE = "weight_data.json"

class WeightEntry:
    def __init__(self, id, date, weight, notes=""):
        self.id = id
        self.date = date
        self.weight = weight
        self.notes = notes

    def to_dict(self):
        return {"id": self.id, "date": self.date, "weight": self.weight, "notes": self.notes}

    @classmethod
    def from_dict(cls, data):
        return cls(data["id"], data["date"], data["weight"], data.get("notes", ""))

class WeightTracker:
    def __init__(self):
        self.entries = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            try:
                with open(DATA_FILE, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.entries = [WeightEntry.from_dict(e) for e in data]
            except:
                self.entries = []

    def save(self):
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump([e.to_dict() for e in self.entries], f, indent=2, ensure_ascii=False)

    def add(self, date, weight, notes=""):
        id = len(self.entries) + 1
        entry = WeightEntry(id, date, weight, notes)
        self.entries.append(entry)
        self.save()
        return id

    def list_all(self):
        if not self.entries:
            print(Fore.YELLOW + "Нет записей.")
            return
        print(Fore.CYAN + f"{'ID':<4} {'Дата':<12} {'Вес (кг)':<12} {'Заметки':<30}")
        print("-" * 60)
        for e in self.entries:
            print(f"{e.id:<4} {e.date:<12} {e.weight:<12.1f} {e.notes:<30}")

    def get_stats(self):
        if not self.entries:
            return None
        weights = [e.weight for e in self.entries]
        dates = [e.date for e in self.entries]
        sorted_entries = sorted(self.entries, key=lambda x: x.date)
        first_weight = sorted_entries[0].weight
        last_weight = sorted_entries[-1].weight
        change = last_weight - first_weight
        # Вычисляем ИМТ (примерно)
        bmi = None
        # Предполагаем рост 175 см для демонстрации
        height_m = 1.75
        if last_weight > 0:
            bmi = last_weight / (height_m * height_m)
        return {
            'count': len(weights),
            'current': last_weight,
            'average': sum(weights) / len(weights),
            'min': min(weights),
            'max': max(weights),
            'change': change,
            'bmi': bmi
        }

    def print_stats(self):
        stats = self.get_stats()
        if not stats:
            print(Fore.YELLOW + "Нет данных для статистики.")
            return
        print(Fore.CYAN + "📊 Статистика:")
        print(f"  Всего записей: {stats['count']}")
        print(f"  Текущий вес: {stats['current']:.1f} кг")
        print(f"  Средний вес: {stats['average']:.1f} кг")
        print(f"  Минимальный вес: {stats['min']:.1f} кг")
        print(f"  Максимальный вес: {stats['max']:.1f} кг")
        change_color = Fore.GREEN if stats['change'] < 0 else Fore.RED if stats['change'] > 0 else Fore.YELLOW
        print(f"  Изменение: {change_color}{stats['change']:+.1f} кг{Style.RESET_ALL}")
        if stats['bmi']:
            bmi_color = Fore.GREEN if 18.5 <= stats['bmi'] <= 24.9 else Fore.YELLOW if 25 <= stats['bmi'] <= 29.9 else Fore.RED
            print(f"  ИМТ: {bmi_color}{stats['bmi']:.1f}{Style.RESET_ALL}")

    def export_csv(self, filename="weight_data.csv"):
        if not self.entries:
            print(Fore.YELLOW + "Нет данных для экспорта.")
            return
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["ID", "Дата", "Вес (кг)", "Заметки"])
            for e in self.entries:
                writer.writerow([e.id, e.date, e.weight, e.notes])
        print(Fore.GREEN + f"💾 Экспорт CSV: {filename}")

    def export_json(self, filename="weight_data_export.json"):
        if not self.entries:
            print(Fore.YELLOW + "Нет данных для экспорта.")
            return
        data = [e.to_dict() for e in self.entries]
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(Fore.GREEN + f"💾 Экспорт JSON: {filename}")

    def delete(self, id):
        for i, e in enumerate(self.entries):
            if e.id == id:
                del self.entries[i]
                self.save()
                return True
        return False

    def edit(self, id, field, value):
        for e in self.entries:
            if e.id == id:
                if field == "date":
                    e.date = value
                elif field == "weight":
                    e.weight = float(value)
                elif field == "notes":
                    e.notes = value
                else:
                    return False
                self.save()
                return True
        return False

    def filter_by_date(self, start_date, end_date):
        results = [e for e in self.entries if start_date <= e.date <= end_date]
        if not results:
            print(Fore.YELLOW + "Нет записей за указанный период.")
            return
        for e in results:
            print(f"{e.id}: {e.date} | {e.weight:.1f} кг | {e.notes}")

def main():
    tracker = WeightTracker()
    while True:
        print(Fore.CYAN + "\n⚖️ Weight Tracker Pro (Python)")
        print("1. Добавить запись")
        print("2. Показать все записи")
        print("3. Статистика")
        print("4. Экспорт в CSV")
        print("5. Экспорт в JSON")
        print("6. Удалить запись")
        print("7. Редактировать запись")
        print("8. Фильтр по дате")
        print("9. Выход")
        choice = input("Выберите действие: ").strip()
        if choice == "1":
            date = input("Дата (ГГГГ-ММ-ДД, Enter для сегодня): ").strip()
            if not date:
                date = datetime.now().strftime("%Y-%m-%d")
            weight = float(input("Вес (кг): "))
            notes = input("Заметки (опционально): ")
            id = tracker.add(date, weight, notes)
            print(Fore.GREEN + f"✅ Запись добавлена (ID: {id})")
        elif choice == "2":
            tracker.list_all()
        elif choice == "3":
            tracker.print_stats()
        elif choice == "4":
            filename = input("Имя CSV файла (по умолч. weight_data.csv): ").strip()
            if not filename:
                filename = "weight_data.csv"
            tracker.export_csv(filename)
        elif choice == "5":
            filename = input("Имя JSON файла (по умолч. weight_data_export.json): ").strip()
            if not filename:
                filename = "weight_data_export.json"
            tracker.export_json(filename)
        elif choice == "6":
            tracker.list_all()
            id = int(input("Введите ID для удаления: "))
            if tracker.delete(id):
                print(Fore.GREEN + "✅ Запись удалена.")
            else:
                print(Fore.RED + "❌ Запись не найдена.")
        elif choice == "7":
            tracker.list_all()
            id = int(input("Введите ID для редактирования: "))
            field = input("Какое поле редактировать (date, weight, notes): ").lower()
            value = input("Новое значение: ")
            if tracker.edit(id, field, value):
                print(Fore.GREEN + "✅ Запись обновлена.")
            else:
                print(Fore.RED + "❌ Не удалось обновить.")
        elif choice == "8":
            start = input("Начальная дата (ГГГГ-ММ-ДД): ")
            end = input("Конечная дата (ГГГГ-ММ-ДД): ")
            tracker.filter_by_date(start, end)
        elif choice == "9":
            print("До свидания!")
            break
        else:
            print(Fore.RED + "Неверный выбор.")

if __name__ == "__main__":
    main()
